Fix invalid DNS-1123 on ACME computed names

meyskens opened the pull request:
**What this PR does / why we need it**:

This is a follow up on https://github.com/jetstack/cert-manager/issues/2505 
While this fixed the CR names we did not ensure this logic on all ACME resources which do the same, causing yet another edge case where this can happen. This doubles down on this check in other generated names. 

**Which issue this PR fixes**:
Fixes https://kubernetes.slack.com/archives/C4NV3DWUC/p1598556555196200

Fixes #3236 

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
Fixes generation of ACME resources if the the 52nd character in a CR name is a symbol.
```

wallrj reviewed (changes requested):
I spotted a typo.

But it'd also be helpful to update the comments explaining or linking to the name requirements.

I searched Kubernetes sources looking how they handle the `meta.generateName` field; all I could find was:

https://github.com/kubernetes/kubernetes/blob/24ec90d23ceec048c8d5a695b4b1b65ce1942b61/staging/src/k8s.io/apiserver/pkg/registry/rest/create.go#L91-L93

And https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/names/generate.go

They don't appear to take any steps to truncate at a non-symbol, but they don't use a `-` separator.

The original reporter @alanjcastonguay in https://github.com/jetstack/cert-manager/issues/2505 uses an example of:

```
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa.example.com
spec:
  secretName: example-com-tls
  dnsNames:
  - example.com
  - www.example.com
  issuerRef:
    name: ca-issuer
    kind: Issuer
    group: cert-manager.io

```

Well when I apply that resource using cert-manager v1-beta.1 I get the following CR name:

```
kubectl get cr -o wide
NAME                                                              READY   ISSUER      STATUS                                                                        AGE
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa.examplnk462   False   ca-issuer   Referenced "Issuer" not found: issuer.cert-manager.io "ca-issuer" not found   4m3s

```

I may be misunderstanding this code, but in that case please add a few comments explaining the desired behaviour here.

meyskens commented:
Seems we broke this in the new cert controllers...

meyskens commented:
@wallrj so the hyphens on GenerateName are added in our code. In the new controllers we moved to GenerateName which removed that util... I added a small check back to solve this for now. On other ACME resources we still use our own (legacy) logic which is kept in this PR but added a general helper function.

meyskens commented:
/test pull-cert-manager-e2e-v1-18

wallrj commented on pkg/api/util/names.go:
```diff
@@ -22,30 +22,33 @@ import (
 
-func ComputeCertificateRequestName(crt *cmapi.Certificate) (string, error) {
-	specBytes, err := json.Marshal(crt.Spec)
+func ComputeName(prefix string, obj interface{}) (string, error) {
```
Add a comment to this function explaining the name requirements and linking to an explanation of the name length limits.

My concern is that the DNSSafeShortenTo52Characters stuff may not be necessary and I want to understand the motivation for it.
It seems to me the the name limit is actually 253 chars, as described here: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names

The DNS label length limit only applies to resources where the name MUST be a valid DNS label, rather than a DNS domain name.

munnerz replied:
These resources are used as the prefix for creating Service, Ingress and Pod resources in the case of ACME. This is where the 63 char limit was 'inherited' from.

wallrj replied:
In that case, perhaps we should apply the truncation when we create those service names and ingress names rather than when creating the CertificateRequest, Order names.

Happy to add a `lgtm` to this now that I understand the reasons, but @meyskens I think we should add a comment or two, recording what @munnerz explained above.

wallrj reviewed (commented):
I just want to check with you whether the 64 char name limit is necessary. See comments below.

wallrj reviewed (approved):
Thanks @meyskens 

/lgtm

meyskens commented:
/test pull-cert-manager-e2e-v1-18

meyskens commented:
/kind bug
